src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,456 @@
+            {
+                nextCommandHandler();
+                return;
+            }
+
+            var root = document.GetSyntaxRootSynchronously(executionContext.OperationContext.UserCancellationToken);
+            var caretPosition = caret.Value.Position;
+
+            var token = GetToken(root, caretPosition, caret.Value);
+
+            var currentNode = token.Parent;
+            // if cursor is right before an opening delimiter, make sure you start with node outside of delimiters
+            if (token.IsKind(SyntaxKind.OpenBraceToken, SyntaxKind.OpenBracketToken, SyntaxKind.OpenParenToken)
+                && token.Span.Start >= caretPosition)
+            {
+                currentNode = currentNode.Parent;
+            }
+
+            if (currentNode == null)
+            {
+                nextCommandHandler();
+                return;
+            }
+
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            if (GetEnclosingArgumentList(currentNode, syntaxFacts) == null)
+            {
+                nextCommandHandler();
+                return;
+            }
+
+            var lastDelimiterSpan = default(TextSpan);
+            var finalDelimiterNeedsSemicolon = false;
+
+            // work your way out, verifying all delimeters exist until you reach statement syntax that requires a semicolon
+            while (!ReachedSemicolonSyntax(currentNode, syntaxFacts, ref finalDelimiterNeedsSemicolon))
+            {
+                if (!ClosingDelimiterExistsIfNeeded(currentNode, ref lastDelimiterSpan))
+                {
+                    // A required delimiter is missing; do not treat semicolon as statement completion
+                    nextCommandHandler();
+                    return;
+                }
+
+                if (currentNode.Parent == null)
+                {
+                    nextCommandHandler();
+                    return;
+                }
+
+                currentNode = currentNode.Parent;
+            }
+
+            // if the statement syntax itself requires a closing delimeter, verify it is there
+            if (finalDelimiterNeedsSemicolon)
+            {
+                if (!StatementClosingDelimiterExists(currentNode, ref lastDelimiterSpan))
+                {
+                    // Example: missing final `)` in `do { } while (x$$`
+                    nextCommandHandler();
+                    return;
+                }
+            }
+
+            // if you haven't found any enclosures, put semicolon at end of statement
+            if (lastDelimiterSpan == default)
+            {
+                lastDelimiterSpan = currentNode.Span;
+            }
+
+            // Move to space after the last delimiter
+            args.TextView.TryMoveCaretToAndEnsureVisible(args.SubjectBuffer.CurrentSnapshot.GetPoint(GetEndPosition(root, lastDelimiterSpan.End, currentNode.Kind())));
+            nextCommandHandler();
```

Comment:
am i wrong... or does every flow path call nextCommandHandler as the very last thing it does.  It feels like we coul dmake this much simpler by pulling out 'nextCommandHandler' to a parent entrypoint and calling into a helper function that never has to deal with it.